File: frontend/src/services/deadlineExtensionService.js
import apiClient from './apiClient';
import { API_PATHS } from '../constants/apiPaths';

/**
 * Service module for admin deadline-extension API calls.
 */

/** Admin: grants (or updates) a deadline extension for a user + type + date. */
export const grantExtension = (payload) => {
  return apiClient.post(API_PATHS.DEADLINE_EXTENSIONS, payload);
};

/** Admin: page of all granted extensions, newest target date first. */
export const getExtensions = (page = 0, size = 20) => {
  return apiClient.get(API_PATHS.DEADLINE_EXTENSIONS, { params: { page, size } });
};

/** Admin: revokes an extension by id. */
export const revokeExtension = (id) => {
  return apiClient.delete(`${API_PATHS.DEADLINE_EXTENSIONS}/${id}`);
};

export const deadlineExtensionService = {
  grantExtension,
  getExtensions,
  revokeExtension,
};

File: src/main/java/com/institute/workforce_tracking/service/DeadlineExtensionService.java
package com.institute.workforce_tracking.service;

import com.institute.workforce_tracking.dto.request.GrantDeadlineExtensionRequest;
import com.institute.workforce_tracking.dto.response.DeadlineExtensionResponse;
import com.institute.workforce_tracking.dto.response.PagedResponse;

/**
 * Admin-granted grace periods on submission deadlines (work plan, work
 * report, lecture summary). Granting also reverses an already-applied
 * penalty for that user and day.
 */
public interface DeadlineExtensionService {

    /**
     * Grants (or updates) an extension. Reverses an existing penalty:
     * un-marks a report absence, restores a SUMMARY_MISSED lecture to
     * COMPLETED, clears a work plan's late flag. Notifies the user.
     */
    DeadlineExtensionResponse grantExtension(String adminEmail,
                                             GrantDeadlineExtensionRequest request);

    /** A page of all granted extensions, newest target date first. */
    PagedResponse<DeadlineExtensionResponse> getExtensions(int page, int size);

    /** Revokes an extension. Does not re-apply penalties retroactively. */
    void revokeExtension(Long extensionId);
}
